Refactor: Enhance movie data handling and caching

skaik-mo opened the pull request:
This commit introduces several improvements to movie data management:

- **Language Change Synchronization:** When the app language changes, movies (except recently viewed) are cleared from the cache, and recently viewed movies are synchronized to reflect the new language.
- **Local Data Fetching:** Use cases for popular, recently released, upcoming, and "matches your vibe" movies now prioritize fetching data from the local cache. Remote fetching occurs only if local data is unavailable.
- **`SyncRecentlyViewedMoviesUseCase`:** A new use case is added to synchronize recently viewed movies, ensuring data consistency across language changes.
- **Repository and DataSource Updates:**
    - Movie repository methods now return `Flow` for local data fetching, enabling reactive updates.
    - Separate methods for local and remote data fetching are introduced for clarity.
    - Local data sources now support adding individual movies and retrieving all recently viewed movies.
- **ViewModel Adjustments:** HomeViewModel now uses `safeCollect` for observing local movie data flows, updating the UI reactively.
- **ShowMoreStrategy Update:** Adjusted to use the new remote fetching methods in use cases.
- **Mapper Enhancement:** Added `toCacheDto()` to `MovieDto` for easier conversion.

skaik-mo committed:
Refactor: Enhance movie data handling and caching

This commit introduces several improvements to movie data management:

- **Language Change Synchronization:** When the app language changes, movies (except recently viewed) are cleared from the cache, and recently viewed movies are synchronized to reflect the new language.
- **Local Data Fetching:** Use cases for popular, recently released, upcoming, and "matches your vibe" movies now prioritize fetching data from the local cache. Remote fetching occurs only if local data is unavailable.
- **`SyncRecentlyViewedMoviesUseCase`:** A new use case is added to synchronize recently viewed movies, ensuring data consistency across language changes.
- **Repository and DataSource Updates:**
    - Movie repository methods now return `Flow` for local data fetching, enabling reactive updates.
    - Separate methods for local and remote data fetching are introduced for clarity.
    - Local data sources now support adding individual movies and retrieving all recently viewed movies.
- **ViewModel Adjustments:** HomeViewModel now uses `safeCollect` for observing local movie data flows, updating the UI reactively.
- **ShowMoreStrategy Update:** Adjusted to use the new remote fetching methods in use cases.
- **Mapper Enhancement:** Added `toCacheDto()` to `MovieDto` for easier conversion.

skaik-mo committed:
Merge branch 'develop' into fix/cache-language

* develop: (68 commits)
  fix safe collect in profile
  add test for Get Movies By Genre Ids Use Case
  fix issues in based on true events and cinematic masterpieces
  fix base view model safe collect
  refactor collection list section ui on click
  Update domain/media/src/test/kotlin/com/giraffe/media/collections/usecase/ClearCollectionsCacheUseCaseTest.kt
  add six strategy to featured collection
  fix test coverage verification
  handle clear collections after logout
  create use case to get movies and series by sort
  create use case to get movies by keywords id
  create use case to get series by keywords id
  delete get filtered use case
  delete discover series function in repo, and add getSeriesByKeywordsId, getSeriesBySort
  delete get filtered use case
  delete discover movies function in repo, and add getMoviesByKeywordsId, getMoviesBySort
  handle clear collections after logout
  refactor late night strategy
  handle getting collections as flow
  - create use case to get series by list of genres - create use case to get series by list of genres - remove use case filter series - remove use case filter movies
  ...

# Conflicts:
#	presentation/home/src/main/java/com/giraffe/presentation/home/screen/show_more/ShowMoreStrategy.kt
#	repository/media/src/main/java/com/giraffe/media/movie/MovieRepositoryImpl.kt

skaik-mo committed:
Refactor: Update movie use case tests for local/remote data sources

This commit updates the following movie use case tests to differentiate between local and remote data sources:

- GetUpcomingMoviesUseCaseTest
- GetPopularityMoviesUseCaseTest
- GetMatchesYourVibeMoviesUseCaseTest
- GetRecentlyReleasedMoviesUseCaseTest

The tests now verify that the correct repository methods (e.g., `getRemoteUpcoming`, `getLocalUpcoming`) are called and that the appropriate data (empty list or fake movies) is returned based on the source.

skaik-mo committed:
Refactor: Observe movie genres as Flow

This commit refactors the way movie genres are fetched and observed throughout the application.

Key changes:
- `MovieRepository.getGenres()` and `getGenresByIds()` now return `Flow<List<Genre>>` instead of `List<Genre>`.
- ViewModels now use `safeCollect` to observe genre changes from the repository.
- A new `SyncMoviesGenresUseCase` is introduced to explicitly trigger a refresh of genres from the remote source. This is used in settings when clearing the cache.
- Local data source now updates existing genres' names if they already exist, instead of just replacing them. This helps preserve the `rank` attribute.
- Mappers and local data source interfaces were updated to reflect these changes.
- Unit tests for genre-related use cases were updated.

dev-bilal-azzam commented on domain/media/src/main/java/com/giraffe/media/movie/repository/MovieRepository.kt:
```diff
@@ -37,16 +37,24 @@ interface MovieRepository {
+
+    suspend fun getRemoteUpcoming(page: Int, limit: Int): List<Movie>
+
+    fun getLocalMatchesYourVibe(limit: Int): Flow<List<Movie>>
```
why you are exposing get Local + get Remote they could be in one function and the repo is responsible when to use which

mohannadahmed00 replied:
> why you are exposing get Local + get Remote they could be in one function and the repo is responsible when to use which

getLocalUpcoming means that he will observe on the movies flow, but getRemoteUpcoming means that he will add new movies form api and then save it in Room so the the flow that collected in home screen will be updated. but i still sthink that names should be changed

dev-bilal-azzam commented on domain/media/src/main/java/com/giraffe/media/movie/usecase/GetMatchesYourVibeMoviesUseCase.kt:
```diff
@@ -6,6 +6,9 @@ import javax.inject.Inject
 ) {
-    suspend operator fun invoke(page: Int, limit: Int = 10) =
-        movieRepository.getMatchesYourVibe(page, limit)
+    fun getLocalMatchesYourVibe(limit: Int = 10) =
```
I think repo should handle which datasource to use not the usecase

dev-bilal-azzam commented on domain/media/src/main/java/com/giraffe/media/movie/usecase/GetPopularMoviesUseCase.kt:
```diff
@@ -6,6 +6,6 @@ import javax.inject.Inject
-    suspend operator fun invoke(page: Int = 1, limit: Int = 10) =
-        movieRepository.getPopular(page = page, limit = limit)
+    operator fun invoke(limit: Int = 10) =
+        movieRepository.getLocalPopular(limit = limit)
```
the same

dev-bilal-azzam commented on domain/media/src/main/java/com/giraffe/media/movie/usecase/GetRecentlyReleasedMoviesUseCase.kt:
```diff
@@ -6,6 +6,9 @@ import javax.inject.Inject
 ) {
-    suspend operator fun invoke(page: Int, limit: Int = 10) =
-        movieRepository.getRecentlyReleased(page = page, limit = limit)
+    fun getLocalRecentlyReleased(limit: Int = 10) =
```
the same

skaik-mo committed:
Merge branch 'develop' into fix/cache-language

* develop:
  Refactor: Simplify safeFlow implementation
  Refactor: Update recently viewed media use cases to support pagination
  Refactor: Wrap SearchRepositoryImpl local calls in safeCall
  Feat: Limit retries for search cache cleanup worker
  Refactor: Improve error handling in SearchRepositoryImpl
  Reduce SearchCacheCleanupWorker frequency
  Refactor: Rename explore module to search
  Refactor: Remove unused search keyword cleaner and use case

skaik-mo committed:
Refactor: Rename movie use cases and repository methods

This commit refactors several use cases and repository methods related to movies:

- Renamed `SyncMoviesGenresUseCase` to `RefreshMoviesGenresUseCase` and updated its corresponding repository method `getRemoteGenres` to `refreshGenres`.
- Renamed `SyncRecentlyViewedMoviesUseCase` to `RefreshRecentlyViewedMoviesUseCase` and updated its corresponding repository method `syncRecentlyViewedMovies` to `refreshRecentlyViewedMovies`.
- Updated various `get...MoviesUseCase` classes to use a single `invoke` method instead of separate `getLocal...` and `getRemote...` methods.
- Correspondingly, repository methods like `getLocalPopular`, `getRemoteRecentlyReleased`, etc., were consolidated into single methods like `getPopular`, `getRecentlyReleased`, etc. These repository methods now handle fetching from local cache or remote API based on the page number.
- Updated calls to these use cases and repository methods across the codebase, including in `SettingsViewModel`, `HomeViewModel`, and various show more strategies.
- Adjusted tests to reflect these changes.

skaik-mo committed:
Refactor: Standardize Movie UseCase Naming and Return Types

This commit refactors several movie use cases to:

- Standardize naming conventions:
    - Renamed `SyncRecentlyViewedMoviesUseCase` to `RefreshRecentlyViewedMoviesUseCase`.
- Ensure consistent return types:
    - Modified use cases that previously returned `suspend fun` to return `Flow` where appropriate (e.g., `GetMatchesYourVibeMoviesUseCase`, `GetUpcomingMoviesUseCase`, `GetRecentlyReleasedMoviesUseCase`).
    - Adjusted use cases to return `Unit` when no specific data is returned (e.g., `RefreshMoviesGenresUseCase`).
- Added unit tests for:
    - `GetMoviesGenresByIdsUseCase`
    - `GetMoviesBySortUseCase`
    - `GetMoviesByKeywordsIdUseCase`
    - `RefreshMoviesGenresUseCase`
    - `RefreshRecentlyViewedMoviesUseCase`
- Updated existing tests to align with the changes.
- In `MovieRepositoryImpl`, ensured `getGenresByIds` returns `emptyFlow()` for empty genre ID lists.
